File: learnLeetCode/Day37/Part2/burst_balloons.cpp
#include <bits/stdc++.h>
using namespace std;

int max_coins(vector<int> nums){
    vector<int> val; val.reserve(nums.size()+2);
    val.push_back(1); for(int x: nums) val.push_back(x); val.push_back(1);
    int n = (int)val.size();
    vector<vector<int>> dp(n, vector<int>(n,0));
    for(int len=2; len<n; ++len){
        for(int l=0; l+len<n; ++l){
            int r=l+len; int best=0;
            for(int k=l+1; k<r; ++k){
                best = max(best, dp[l][k] + val[l]*val[k]*val[r] + dp[k][r]);
            }
            dp[l][r] = best;
        }
    }
    return dp[0][n-1];
}

int main(){ ios::sync_with_stdio(false); cin.tie(nullptr);
    cout << max_coins(vector<int>{3,1,5,8}) << "\n"; // Expected 167
    return 0; }

File: learnLeetCode_remote/Day37/Part2/burst_balloons.py
from typing import List

def max_coins(nums: List[int]) -> int:
    val = [1] + nums + [1]
    n = len(val)
    dp = [[0]*n for _ in range(n)]
    for len_ in range(2, n):  # interval size
        for l in range(0, n - len_):
            r = l + len_
            best = 0
            for k in range(l+1, r):
                best = max(best, dp[l][k] + val[l]*val[k]*val[r] + dp[k][r])
            dp[l][r] = best
    return dp[0][n-1]

if __name__ == "__main__":
    print(max_coins([3,1,5,8]))  # Expected 167
